### ISSUE TITLE
Major change proposal: Support collecting all identifiers encountered during compilation

### ISSUE THREAD

**Issue by crlf0710** [rust-lang/compiler-team#276]
# TL;DR
To implement `confusable_idents` lint in RFC 2457, we need a mechanism to collect all unique identifiers occurred during compilation.
# Links and Details
The requirement raised up from [RFC 2457](https://github.com/rust-lang/rfcs/blob/master/text/2457-non-ascii-idents.md#confusable-detection). Fundamentally, it means that each occurred identifier should be hashed with a very specific hash function called `skeleton` and storing the results. The lint is about checking whether hash collisions occurred during compilation. Whenever that happens, it raises warnings or errors.

# Mentors or Reviewers
(TBD)

**Comment by spastorino** [comment]
Added to our compiler team meeting agenda, removing `to-announce` label.

**Comment by crlf0710** [comment]
Done and merged in https://github.com/rust-lang/rust/pull/71542 , i'm gonna close this.
